pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,279 @@
+	}
+	isEmpty, err := isDirEmpty(path)
+	if err != nil {
+		return false, err
+	}
+	if !isEmpty {
+		return true, nil
+	}
+	return false, nil
+}
+
+// getChartNameAndVersion receives a filepath and returns the chart name and chart version.
+// The path has to follow /{root}/{chartName}/{chartVersion}/*
+func getChartNameAndVersion(file string) (string, string, error) {
+	root, err := filepath.Abs(FSCacheRootDir)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get absolute root path: %w", err)
+	}
+	filePath, err := filepath.Abs(file)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get absolute file path: %w", err)
+	}
+
+	// file has to be rooted at FSCacheRootDir
+	if strings.HasPrefix(filePath, root) {
+		p, _ := strings.CutPrefix(filePath, root)
+		s := strings.Split(p, string(os.PathSeparator))
+		if len(s) < 3 {
+			return "", "", fmt.Errorf("file path is not valid. Path provided: %s", file)
+		}
+		chartName := s[1]
+		chartVersion := s[2]
+		if _, err := semver.NewVersion(chartVersion); err != nil {
+			return "", "", fmt.Errorf("invalid chart version: %w", err)
+		}
+		return chartName, chartVersion, nil
+	}
+	return "", "", fmt.Errorf("path root is not the root cache path. Path provided: %s", file)
+
+}
+
+func fsCacheFilepathGlob(pattern string) ([]string, error) {
+	files, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debugf("files matching glob pattern [%s] found in filesystem cache: %+v", pattern, files)
+
+	return files, nil
+}
+
+// getVersionFromPackageJSON takes in a URL for a plugin's package.json, reads it, and returns a Semver object of the version contained in the file
+func getVersionFromPackageJSON(packageJSONURL string) (*semver.Version, error) {
+	data, err := readFile(packageJSONURL)
+	if err != nil {
+		return nil, err
+	}
+	var packageJSON PackageJSON
+	err = json.Unmarshal(data, &packageJSON)
+	if err != nil {
+		return nil, err
+	}
+	version, err := semver.NewVersion(packageJSON.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	return version, nil
+}
+
+// readFilesTxt takes in a URL for a plugin's files.txt, reads it, and returns a slice of the file paths contained in the file
+func readFilesTxt(filesTxtURL string) ([]string, error) {
+	data, err := readFile(filesTxtURL)
+	if err != nil {
+		return nil, err
+	}
+	files := strings.Split(string(data), "\n")
+
+	return files, nil
+}
+
+// readFile reads the file from the given URL and returns the data
+func readFile(URL string) ([]byte, error) {
+	logrus.Debugf("fetching file [%s]", URL)
+	resp, err := http.Get(URL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	maxFileSize, err := strconv.ParseInt(settings.MaxUIPluginFileByteSize.Get(), 10, 64)
+	if err != nil {
+		logrus.Errorf("failed to convert setting MaxUIPluginFileByteSize to int64, using fallback. err: %s", err.Error())
+		maxFileSize = 20 << 20 //20mb
+	}
+	reader := &io.LimitedReader{R: resp.Body, N: maxFileSize + 1}
```

Comment:
Good web servers would tell us the correct amount of data that is being returned, so we can short-circuit the process by validating that first:

```suggestion
	if resp.ContentLength > maxFileSize {
		return nil, errMaxFileSizeError
	}
	
	reader := &io.LimitedReader{R: resp.Body, N: maxFileSize + 1}
```

The `LimitedReader` would then be used as a secondary measure, in case the web server tells us one thing, but ends up providing us more data then it previously promised.

Reply:
Done